feat(search): s2 migration

mdt2 opened the pull request:
## Description

I believe this is blocked until we get the S2 icons.

S2 migration of Search to align with [design specs](https://www.figma.com/file/eoZHKJH9a3LJkHYCGt60Vb/S2-token-specs?type=design&node-id=10954-4435&mode=design&t=Hh1ne7NyJR877hSV-0). On top of the migration work, this PR includes:

- Removal of /themes
- Removal of quiet variant
- Increased Chromatic coverage
- Minor mod adjustments
- Updated migration guide notes

## How and where has this been tested?

Please tag yourself on the tests you've marked complete to confirm the tests have been run by someone other than the author.

### Validation steps

- [ ] [Storybook testing preview](https://pr-2673--spectrum-css.netlify.app/preview/?path=/story/components-search--default&globals=testingPreview:!true) shows all variants (except for hover as we cant show that without adding a class)
- [ ] Test WHCM in light and dark mode
- [ ] Be sure to test hover while focused since there are specific tokens applied for this
- [ ] Design validation

### Regression testing

Validate:

1. The documentation pages for at least two other components are still loading, including:

- [ ] The pages render correctly, are accessible, and are responsive.

2. If components have been modified, VRTs have been run on this branch:

- [ ] VRTs have been run and looked at.
- [ ] Any VRT changes have been accepted (by reviewer and/or PR author), or there are no changes.

## To-do list

- [x] I have read the [contribution guidelines](/.github/CONTRIBUTING.md).
- [x] I have updated relevant storybook stories and templates.
- [x] I have tested these changes in Windows High Contrast mode.
- [x] If my change impacts **documentation**, I have updated the documentation accordingly.
- [ ] ✨ This pull request is ready to merge. ✨

mdt2 commented on components/search/stories/search.stories.js:
```diff
@@ -16,29 +19,37 @@ export default {
 		},
-		isDisabled: {
-			name: "Disabled",
+		isFocused: {
+			table: { disable: true },
+			type: { name: "boolean" },
+		},
```
I added this because TextField uses this class in it's Storybook, so we can use this arg to [pass through to TextField](https://github.com/adobe/spectrum-css/pull/2673/files#diff-5a941796927ca4c2dd96d02d4b51ac05e7f1a0c2d9a7ba3292cf2948a67245baR33) to show the focus state in our Chromatic testing preview for Search.

rise-erpelding commented on components/search/index.css:
```diff
@@ -298,6 +325,18 @@ governing permissions and limitations under the License.
 			var(--mod-search-border-width, var(--spectrum-search-border-width))
 	);
+	padding-inline-start: calc(
+		var(--mod-search-edge-to-text, var(--spectrum-search-edge-to-text)) -
```
Is there a reason for using `edge-to-text` rather than `edge-to-visual` here?

mdt2 replied:
Great catch, I missed this one.

rise-erpelding reviewed (changes requested):
The CSS for this is looking so much cleaner and the testing preview matches several of the others that we've done! ✨ Here are a few visual things I noticed when going through this PR that might need some changes:

* The border for the disabled state is transparent but should have a gray color, [it looks like there wasn't a `--mod` for `border-color` ever set on the disabled state for textfield](https://github.com/adobe/spectrum-css/blob/main/components/textfield/index.css#L1099):
<img width="237" alt="image" src="https://github.com/adobe/spectrum-css/assets/54716846/9473f5bd-cf58-4243-9f57-32e32a4fc908">

* I'm wondering if this might need an `.is-keyboardFocused:hover` selector, this hover state looks like it might be unintentional (it goes gray from black): 
<img width="234" alt="image" src="https://github.com/adobe/spectrum-css/assets/54716846/e5d9480a-45e4-4c8e-87e4-e94965a18591">

**Other questions I have that were carryover from how the component was before:**
* Is there a way to preview the search field with the help text? I know there's a spec for spacing the help text from the component, and I see that it's being used in the code, but I didn't see a way to view it in Storybook.
* Would it be helpful at all to have a control to show/not show the clear button? I was thinking if it wasn't showing by default I could check the `pill-edge-to-text` spacing but then I realized that anytime it'd be used, the text would always be the placeholder "Search" text and the spacing between the end of that text box and the edge of the text field is a bit irrelevant. But is there any reason why anyone might want to look at the search field without the clear button instead of it always being there?
* In WHCM, there's a clear color difference between the text color of the placeholder text and the text color when there's a value, it was like this before, and it's probably not a big deal, but it seemed a little odd to me so I thought I'd call it out:
![image](https://github.com/adobe/spectrum-css/assets/54716846/5349bec2-7c36-4fad-a919-ed9b310175cd)
![image](https://github.com/adobe/spectrum-css/assets/54716846/1e34642c-f3af-4865-97dd-615d6f5c6085)

jawinn reviewed (commented):
One thing I am wondering about is the focus and focus indicator on the clear button. In our Storybook, you can tab focus the field and then tab focus the clear button, which shows a rounded focus indicator.

I didn't see a design for that focus on the clear button to reference, and also took a look at the [SWC search](https://opensource.adobe.com/spectrum-web-components/storybook/index.html?path=/story/search-sizes--m), which doesn't have it tab focusable (the field clears with the Esc key it seems). Is the intention for us to not have it as a tab stop?

mdt2 commented on components/search/stories/search.stories.js:
Great idea!

mdt2 commented:
Thanks for the feedback! Some responses for Rise:

- **Disabled border color** - Great callout. I've updated Textfield to improve this which will probably just be a patch change.
- **.is-keyboardFocused:hover** - I've also added this, good catch.
- **Help text** - Yeah totally! I added some help text sections to our Chromatic coverage, and you should see a control to show the help text and also the ability to change the text now.
- **Clear button** - I went down a bit of a rabbit hole with this one trying to use `useArgs()` to only show this button when there's an input value. I wasn't able to get it working within a reasonable timebox and came to the same conclusion you did about the spacing.
- **WHCM placeholder color** - I noticed this too. I think that using the disabled color for the placeholder text can help communicate that its placeholder text, so that seemed ok to me. Let me know if you have a different value you think would work better!

And a response for Josh:

- **Clear button focus** - This feels like it's in that gray area of whether we would handle this or if it's an implementation detail 🤔 Seems like removing the focus on our end would be a Storybook-only adjustment and would look something like adding an arg to ClearButton to adjust aria-hidden and tab-index, is that what you were thinking? I can add a commit for that and see what you think.

castastrophe commented on components/search/index.css:
```diff
@@ -26,49 +24,61 @@ governing permissions and limitations under the License.
-		--spectrum-disabled-background-color
-	);
+
+	/* Used to reassign to Textfield mods */
+	/* stylelint-disable-next-line spectrum-tools/no-unused-custom-properties */
+	--spectrum-search-background-color-disabled: var(--spectrum-gray-25);
+	/* stylelint-disable-next-line spectrum-tools/no-unused-custom-properties */
+	--spectrum-search-border-color-disabled: var(--spectrum-disabled-border-color);
```
```suggestion
	/* stylelint-disable-next-line spectrum-tools/no-unused-custom-properties -- reassign to Textfield mods */
	--spectrum-search-background-color-disabled: var(--spectrum-gray-25);
	/* stylelint-disable-next-line spectrum-tools/no-unused-custom-properties -- reassign to Textfield mods */
	--spectrum-search-border-color-disabled: var(--spectrum-disabled-border-color);
```

If you format the note like this, you won't get a warning about descriptionless disables for stylelint.

castastrophe commented on components/search/stories/search.stories.js:
```diff
@@ -1,10 +1,13 @@
 	description:
 		"This component contains a single input field with both a magnifying glass icon and a \"reset\" button displayed within it.",
-	component: "Search",
+	component: "Search field",
```
```suggestion
	component: "SearchField",
```

The component name field needs to be formatted like a React component naming rather than a readable name.

castastrophe commented on components/search/stories/search.stories.js:
```diff
@@ -52,7 +82,174 @@ export default {
+					border: 1px solid var(--spectrum-gray-200);
+					border-radius: 4px;
+					padding: 0 1em 1em;
+					/* Why seafoam? Because it separates it from the component styles. */
```
Ah love to see it! <3

marissahuysentruyt commented on components/clearbutton/stories/clearbutton.stories.js:
```diff
@@ -51,6 +51,11 @@ export default {
 			...staticColor,
 			options: ["white"],
 		},
+		variant: { table: { disable: true }},
```
TODO: not sure what this is- remove.

marissahuysentruyt commented on components/search/index.css:
```diff
@@ -25,49 +35,44 @@
-	--spectrum-search-edge-to-visual: var(--spectrum-component-edge-to-visual-100);
-
-	/* Focus Indicator */
-	--spectrum-search-focus-indicator-thickness: var(--spectrum-focus-indicator-thickness);
-	--spectrum-search-focus-indicator-gap: var(--spectrum-focus-indicator-gap);
-	--spectrum-search-focus-indicator-color: var(--spectrum-focus-indicator-color);
```
Definitely open to keeping these focus indicator properties in. I figured since search field is basically extending textfield, maybe these weren't necessary though.

marissahuysentruyt commented on components/search/stories/template.js:
```diff
@@ -1,3 +1,4 @@
+
```
TODO: remove extra line.

marissahuysentruyt commented on tokens/custom/global-vars.css:
```diff
@@ -61,8 +61,11 @@
 	/* Gradients that do not change with the color theme, for use in static color backgrounds. */
-	--spectrum-examples-gradient-static-black: linear-gradient(45deg, rgb(255, 241, 246), rgb(238, 245, 255));
-	--spectrum-examples-gradient-static-white: linear-gradient(45deg, rgb(64, 0, 22), rgb(14, 24, 67));
+	--spectrum-examples-gradient-static-black: linear-gradient(45deg, rgb(255 241 246), rgb(238 245 255));
+	--spectrum-examples-gradient-static-white: linear-gradient(45deg, rgb(64 0 22), rgb(14 24 67));
```
Not sure why these were changed after the rebase. 

TODO: revert the commas.

marissahuysentruyt commented on tokens/custom/global-vars.css:
```diff
@@ -61,8 +61,11 @@
 	--spectrum-coach-indicator-ring-static-white-color: var(--spectrum-white);
+
+	/* override for Spectrum Tokens value until a Style Dictionary transform gets written for our usage */
+	--spectrum-corner-radius-1000: 9999px;
```
This adds the pill corner rounding as a custom property. I can ask Garth about that current value of `corner-radius-1000`, and dig into our broken corner rounding guidance page.

marissahuysentruyt commented on .changeset/shaggy-schools-sing.md:
```diff
@@ -0,0 +1,15 @@
+---
+"@spectrum-css/search": major
+"@spectrum-css/textfield": patch
```
This may not have to be textfield now, and should be tokens? Since we added the custom corner-radius-1000?

rise-erpelding commented on components/search/stories/search.test.js:
Thank you for the added coverage on this!

rise-erpelding commented on components/search/index.css:
```diff
@@ -193,6 +208,8 @@
 	padding-block-start: calc(var(--mod-search-top-to-text, var(--spectrum-search-top-to-text)) - var(--mod-search-border-width, var(--spectrum-search-border-width)));
 	padding-block-end: calc(var(--mod-search-bottom-to-text, var(--spectrum-search-bottom-to-text)) - var(--mod-search-border-width, var(--spectrum-search-border-width)));
+	padding-inline-start: calc(var(--mod-search-edge-to-visual, var(--spectrum-search-edge-to-visual)) - var(--mod-search-border-width, var(--spectrum-search-border-width)) + var(--mod-search-icon-size, var(--spectrum-search-icon-size)) + var(--mod-search-text-to-icon, var(--spectrum-search-text-to-icon)));
+	padding-inline-end: calc(var(--mod-search-button-inline-size, var(--spectrum-search-button-inline-size)) - var(--mod-search-border-width, var(--spectrum-search-border-width)));
```
Is this the piece of padding between the text input and the edge?

I noticed we're not using `component-pill-edge-to-text` tokens here and I'm wondering why? Is it because we're trying to leave more space for the clear button?

But also the clear button isn't really spec'd inside of search at all, which feels odd to me, is there a reason for that, that you know of?

marissahuysentruyt replied:
Yes- the `padding-inline-start` should add space between the value & placeholder text and the search icon, and then the `padding-inline-end` should give space between the value in the input and the clear button (that's when it's most noticeable I think).

The `--spectrum-search-edge-to-visual` should be set to the component pill edge to text tokens for each t-shirt size. But if you're asking why aren't we using that in the `padding-inline-end`, I don't have a good answer for that. I'll have to get back with you.

I don't know why the clear button isn't spec'd, but I can certainly reach out to design and ask. All I can infer from the designs is that the clear button is all the way at the end of the search container. That's how the component seems to be styled currently as well.

rise-erpelding reviewed (changes requested):
This is mostly a pretty straightforward migration, and doesn't need a whole lot! There are some really nice additions/enhancements to the docs and testing template. The only remnants of the quiet variant in this component are in the changelog, which is expected. My notes:
- We'll want to do a rebase on this to include the textfield migration
- Definitely should clarify what we're ok with carrying over from textfield and what should be separately defined in search.
- Questioning the end spacing on the input, the token not being used there, and the clear button spacing specs.
- There was a small change in spec that we talked about separately where we want the keyfocus border color to be gray-800 instead of gray-900 to align with the colors of textfield... not a huge deal and we _can_ wait for the "official" spec to change and follow up later if you think that's a better idea.
- Last thing - what is the minimised variant? I assume it's just like a search icon that you click on and then the search field appears? Do you think we'd need to do anything with that?

jawinn commented on components/search/stories/search.stories.js:
```diff
@@ -13,40 +13,56 @@ import { SearchOptions, Template } from "./template.js";
+		isFocused,
+		isKeyboardFocused,
+		inputValue: {
+			name: "Display value and close button",
+			description: "When defined, this will display the input and the clear button within the search field.",
```
```suggestion
			name: "Value",
			description: "When defined, this will set the value of the input, and the clear button will appear within the search field.",
```
I was a little confused by the title of this one at first (partly because it said close button and not clear button). I'd suggest just calling it "Value".

jawinn replied:
I think it also would be helpful to show an example with a value and the clear button on the Docs page.

jawinn reviewed (changes requested):
Here are a couple more suggestions.

marissahuysentruyt commented on components/search/index.css:
```diff
@@ -25,49 +35,44 @@
+	&:lang(ja),
+	&:lang(zh),
+	&:lang(ko) {
+		--spectrum-search-line-height: var(--spectrum-cjk-line-height-100);
```
Yeah....I might keep the font styles and the focus indicator styles all in. I think Express had concerns around inheriting styles in menu and menu item, and I believe their request was to explicitly define them in each component. I'll see if I can find that Slack thread and send it to you.

It feels redundant, but maybe it's best to keep the repeats in for now.

marissahuysentruyt commented on components/search/index.css:
```diff
@@ -84,54 +89,63 @@
+	display: inline-block;
+	position: relative;
+	inline-size: var(--mod-search-inline-size, var(--spectrum-search-inline-size));
+	min-inline-size: var(--mod-search-min-inline-size, var(--spectrum-search-min-inline-size));
```
Just doing a quick search of the repo for components that offer a `--mod-component-min-inline-size` (or min-width), it looks like combobox, picker, text area, action button, dialog, side nav, and several others all offer something like that. It's not a huge of amount of components- maybe 15 to 20- but it's also not "only a handful." 

I don't disagree that if users _really_ wanted to change the size that they could just use `--mod-search-inline-size`. And I mean, I guess this is a breaking change in the component anyways, so as long as it's documented that the min inline size mod is removed, it should be fine. 🤨 

I don't know that I have a strong enough opinion. What do you think- remove the mod for min inline size?

marissahuysentruyt commented on components/search/stories/search.stories.js:
```diff
@@ -13,40 +13,56 @@ import { SearchOptions, Template } from "./template.js";
+		},
+		isFocused,
+		isKeyboardFocused,
+		inputValue: {
+			name: "Display value and close button",
+			description: "When defined, this will display the input and the clear button within the search field.",
```
Good catch on the `inputValue`- thank you!

For the clear button....yeah that's a tricky one. If I use the `inputValue` controls, the clear button responds as I'm expecting- appearing when that `inputValue` is defined, and not appearing when `inputValue` is undefined. But there's a few issues if I actually use the component, like when I type into the search field itself, the button _doesn't_ appear at all. I think you're right about this aspect of the component is stuck in that "resetting" issue we've found before. 

https://github.com/user-attachments/assets/4bbb1575-c882-4079-936a-e308f7b6dd75

Should I leave an implementation note in the docs about the expected behavior? "Implementations should be sure that the clear button renders when an input value is defined, and does not render when that value is undefined" (or something like that?)

marissahuysentruyt commented on components/search/stories/search.stories.js:
```diff
@@ -56,6 +72,9 @@ export default {
 				"click .spectrum-Search-icon",
 			],
 		},
+		status: {
+			type: "migrated",
+		},
```
I don't see it on the docs page. We definitely had a rendered tag, at least on deprecated components, at one point, but I think this "migrated" tag is just metadata, not indicated visually anywhere. 

There's only a "deprecated" status in preview.js, so I'm not sure this is particularly doing anything currently. Just for funsies, I changed this to "deprecated" and can get the tag. Maybe a migrated tag could be a future feature for us?? If we can get spectrum-two as our main, a migrated tag would be super useful!

<img width="485" alt="Screenshot 2025-04-10 at 8 59 49 AM" src="https://github.com/user-attachments/assets/1cfde8c6-7749-4030-a801-c6e0b738b338" />

marissahuysentruyt commented on components/search/index.css:
```diff
@@ -204,32 +221,20 @@
-		padding-inline-start: calc(var(--mod-search-edge-to-visual, var(--spectrum-search-edge-to-visual)) - var(--mod-search-border-width, var(--spectrum-search-border-width)) + var(--mod-search-icon-size, var(--spectrum-search-icon-size)) + var(--mod-search-text-to-icon, var(--spectrum-search-text-to-icon)));
-		padding-inline-end: var(--mod-search-button-inline-size, var(--spectrum-search-button-inline-size));
+@media (forced-colors: active) {
+	.spectrum-Search .spectrum-Search-textfield,
+	.spectrum-Search .spectrum-Search-textfield .spectrum-Search-input {
+		--highcontrast-search-color-default: CanvasText;
+		--highcontrast-search-color-hover: CanvasText;
+		--highcontrast-search-color-focus: CanvasText;
+		--highcontrast-search-color-disabled: GrayText;
```
@rise-erpelding I've updated this, but didn't see the same colors as your screenshot (neither in the Chrome forced-colors emulator, or in AssistivLabs). 

<img width="288" alt="Screenshot 2025-04-10 at 12 42 06 PM" src="https://github.com/user-attachments/assets/22d15945-63b0-4fee-a36e-98152cc7a102" />

I did notice that the focus+hover was black? I'd like your advice on how to approach that border color change, since the "simplest" fix was just to add the `--highcontrast-textfield-border-color-focus-hover` variable to the stack. Is that the right approach? Search itself didn't have any other highcontrast border custom props, but textfield does. Let me know if you see the same thing or if you have a better idea on how to handle that border color correctly.

<img width="294" alt="Screenshot 2025-04-10 at 12 42 17 PM" src="https://github.com/user-attachments/assets/51cb05a2-a8ab-4a5a-b29b-8d57b28f5a6c" />

marissahuysentruyt commented on .changeset/shaggy-schools-sing.md:
```diff
@@ -0,0 +1,15 @@
+      - Added `--mod-search-edge-to-text`.
+      - Added `--mod-search-top-to-icon`.
+      - Removed `--mod-workflow-icon-size-100`.
+- chore(textfield): add border-color custom property mod
+  - Search needed access to the disabled border-color in textfield, so the `--mod-textfield-border-color-disabled` custom property was added.
```
I'm not sure I'm totally following your question on the search border color. I do see this in all of the passthrough mods: 

`--mod-textfield-border-color-disabled: var(--mod-search-border-color-disabled, var(--spectrum-search-border-color-disabled));`

Does that answer any of your questions? This seems to track with the pattern that was happening in this file, but I haven't come across many other migrations that seem to have done this same, mod-heavy approach. We might have to pair if I'm still not understanding.

marissahuysentruyt commented:
> * Last thing - what is the minimised variant? I assume it's just like a search icon that you click on and then the search field appears? Do you think we'd need to do anything with that?


@rise-erpelding I have CSS-1158 written up for the minimized variant. When I last heard from Kami and Lynn, the file/branch for that variant was corrupted so Kami was in the midst of recreating I believe. The minimized search is basically just a search icon-only button, and when clicked, would expand the field, if I remember correctly. I can share Slack threads in a DM if you'd like!

rise-erpelding commented on components/search/index.css:
```diff
@@ -25,41 +35,45 @@
+	--spectrum-search-top-to-icon: var(--spectrum-component-top-to-workflow-icon-100);
 
 	/* Focus Indicator */
 	--spectrum-search-focus-indicator-thickness: var(--spectrum-focus-indicator-thickness);
```
@marissahuysentruyt I wonder if we need to add a line like 

```/* stylelint-disable-next-line spectrum-tools/no-unused-custom-properties -- used to assign Textfield mods */```

to keep this from making noise.

rise-erpelding commented:
> > * Last thing - what is the minimised variant? I assume it's just like a search icon that you click on and then the search field appears? Do you think we'd need to do anything with that?
> 
> @rise-erpelding I have CSS-1158 written up for the minimized variant. When I last heard from Kami and Lynn, the file/branch for that variant was corrupted so Kami was in the midst of recreating I believe. The minimized search is basically just a search icon-only button, and when clicked, would expand the field, if I remember correctly. I can share Slack threads in a DM if you'd like!

I did see that ticket at the last Backlog review, thanks!! No further details needed for now 🙂

jawinn commented on components/search/index.css:
```diff
@@ -84,54 +98,63 @@
-			forced-color-adjust: none;
-			background-color: transparent;
-		}
+	& > .is-focused.is-hover .spectrum-Search-input {
```
I noticed when looking over the metadata.json that we have a selector using a storybook only class `.is-hover`. I'm guessing that this selector needs to change to use `:hover`.

marissahuysentruyt replied:
@jawinn fantastic catch! https://github.com/adobe/spectrum-css/pull/2673/commits/ecdbc746473720e2886f38572ae25456678ce2d9

rise-erpelding reviewed (approved):
Nothing else blocking for me!

This is minor, but I think the reviewdog stylelint comments might need addressing as I'm seeing those same lines underlined in my editor.

I think we discussed the keyfocus border color change somewhere at some point (we want it to be `gray-800` like the mouse focus, not `gray-900`), I don't see that in here, but that's fine because the border colors match what's currently in the spec. Let's make a note to keep an eye out for a spec change here though since we were told unofficially that the value is wrong and we'll make the change then!

Nice work on this!
